src/rng.cpp

```diff
@@ -7,40 +7,70 @@
 #include <cstdlib>
 #include <random>
 
-long rng( long val1, long val2 )
+static std::mt19937 &rng_get_engine();
+static std::uniform_int_distribution<long> rng_int_dist;
+static std::uniform_real_distribution<double> rng_real_dist;
+static std::normal_distribution<double> rng_normal_dist;
+
+uintmax_t rng_bits()
 {
-    long minVal = ( val1 < val2 ) ? val1 : val2;
-    long maxVal = ( val1 < val2 ) ? val2 : val1;
-    return minVal + long( ( maxVal - minVal + 1 ) * double( rand() / double( RAND_MAX + 1.0 ) ) );
+    size_t size = sizeof( uintmax_t ) * CHAR_BIT;
+    size_t word = rng_get_engine().word_size;
+    size_t shift = 0;
+    div_t res = div( size, word );
+    if( res.rem ) {
+        res.quot++;
+    }
+    res.quot--;
+    uintmax_t ret = 0;
+    while( 1 ) {
+        ret |= uintmax_t( rng_get_engine()() ) << shift;
+        if( res.quot > 0 ) {
+            shift += word;
+            res.quot--;
+            continue;
+        }
+        break;
+    }
```

Comment:
Perhaps simpler as
```
while( res.quot ) {
    ret <<= word;
    ret |= rng_get_engine()();
    --res.quot;
}
```
Then you don't need the `shift` variable at all.

Reply:
Your loop would not execute in the case where `sizeof (uintmax_t) / rng_get_engine().word_size` equals `1` (one could switch to the mt19937 engine, which has a `word_size` of 64, which is probably the same as `sizeof (uintmax_t) * CHAR_BIT`). 

Furthermore, in the same case if the loop were to execute (such as removing the line containing `res.quot--;`, you would be shifting `(uintmax_t) by sizeof (uintmax_t) * CHAR_BIT`, but shifting by more than `sizeof (uintmax_t) * CHAR_BIT - 1` is undefined behaviour.

Reply:
Then perhaps
```
while( true ) {
    ret |= rng_get_engine()();
    if ( res.quot == 0 ) {
        break;
    }
    ret <<= word;
    --res.quot;
}
```